File: test/features/caffeine/caffeine_display_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/core/time/local_date.dart';
import 'package:openvitals/domain/model/caffeine_models.dart';
import 'package:openvitals/features/caffeine/application/caffeine_display.dart';

/// The derivations the caffeine cards used to do in their build paths — the
/// sleep-impact verdict, the top six slices and their shared scale, the curve's
/// axis maximum — now a pure function the view-model calls once per load.
CaffeineDistributionSlice _slice(String label, double mg) =>
    CaffeineDistributionSlice(label: label, valueMg: mg);

void main() {
  final morning = DateTime(2026, 3, 2, 8);

  test('an empty load leaves the bars empty and the scale non-zero', () {
    final display = buildCaffeineDisplay(
      home: const CaffeineInsights(),
      analytics: const CaffeineInsights(),
    );

    expect(display.home.sleepImpactStatus, CaffeineSleepImpactStatus.unlikely);
    expect(display.home.curveEntryTimes, isEmpty);
    // The threshold line and the curve both have to fit in something.
    expect(display.home.curveMaxMg, 1.0);
    expect(display.analytics.sourceBars, isEmpty);
    expect(display.analytics.topSourceLabel, isNull);
  });

  test('the sleep verdict compares bedtime first, then right now', () {
    CaffeineSleepImpactStatus statusFor({
      required double currentMg,
      required double bedtimeMg,
    }) =>
        buildCaffeineDisplay(
          home: CaffeineInsights(
            currentMg: currentMg,
            bedtimeMg: bedtimeMg,
            sleepThresholdMg: 50,
            bedtime: const LocalTime(23, 0),
          ),
          analytics: const CaffeineInsights(),
        ).home.sleepImpactStatus;

    expect(statusFor(currentMg: 120, bedtimeMg: 80),
        CaffeineSleepImpactStatus.mayAffectSleep);
    // Over the line now, but back under it by bedtime.
    expect(statusFor(currentMg: 120, bedtimeMg: 30),
        CaffeineSleepImpactStatus.elevatedNow);
    expect(statusFor(currentMg: 20, bedtimeMg: 5),
        CaffeineSleepImpactStatus.unlikely);
  });

  test('the bedtime card is safe exactly at the threshold', () {
    final display = buildCaffeineDisplay(
      home: const CaffeineInsights(bedtimeMg: 50.0, sleepThresholdMg: 50),
      analytics: const CaffeineInsights(),
    );

    expect(display.home.bedtimeIsSafe, isTrue);
    // …and the verdict agrees: 50 is not ABOVE 50.
    expect(display.home.sleepImpactStatus, CaffeineSleepImpactStatus.unlikely);
  });

  test('the curve maximum fits the tallest point and the threshold', () {
    final display = buildCaffeineDisplay(
      home: CaffeineInsights(
        sleepThresholdMg: 50,
        curvePoints: [
          CaffeinePoint(time: morning, valueMg: 90.0),
          CaffeinePoint(
            time: morning.add(const Duration(hours: 2)),
            valueMg: 140.0,
          ),
        ],
      ),
      analytics: const CaffeineInsights(),
    );

    expect(display.home.curveMaxMg, 140.0);
  });

  test('a distribution card shows six bars, scaled against their own tallest',
      () {
    final display = buildCaffeineDisplay(
      home: const CaffeineInsights(),
      analytics: CaffeineInsights(
        sourceTotals: [
          _slice('Coffee', 400),
          _slice('Tea', 200),
          _slice('Cola', 100),
          _slice('Chocolate', 50),
          _slice('Energy drink', 40),
          _slice('Matcha', 30),
          // The seventh does not fit on the card.
          _slice('Yerba mate', 20),
        ],
        timeBuckets: const [
          CaffeineTimeBucket(
            bucket: CaffeineTimeOfDayBucket.morning,
            valueMg: 300,
          ),
          CaffeineTimeBucket(
            bucket: CaffeineTimeOfDayBucket.evening,
            valueMg: 150,
          ),
        ],
      ),
    );

    final bars = display.analytics.sourceBars;
    expect(bars.length, 6);
    expect(bars.first.label, 'Coffee');
    expect(bars.first.fraction, 1.0);
    expect(bars[1].fraction, 0.5);
    expect(display.analytics.topSourceLabel, 'Coffee');

    final buckets = display.analytics.timeBucketBars;
    expect(buckets.first.bucket, CaffeineTimeOfDayBucket.morning);
    expect(buckets.first.fraction, 1.0);
    expect(buckets.last.fraction, 0.5);
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/features/caffeine/CaffeineDisplayTest.kt
package tech.mmarca.openvitals.features.caffeine

import java.time.Instant
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import tech.mmarca.openvitals.domain.model.CaffeineDistributionSlice
import tech.mmarca.openvitals.domain.model.CaffeineInsights
import tech.mmarca.openvitals.domain.model.CaffeinePoint
import tech.mmarca.openvitals.domain.model.CaffeineTimeBucket
import tech.mmarca.openvitals.domain.model.CaffeineTimeOfDayBucket

/** The caffeine cards' derivations as pure functions: the sleep verdict, the top six slices, the curve's axis maximum. */
class CaffeineDisplayTest {

    private val morning: Instant =
        LocalDateTime.of(2026, 3, 2, 8, 0).atZone(ZoneId.systemDefault()).toInstant()

    @Test
    fun `an empty load leaves the bars empty and the scale non-zero`() {
        val home = CaffeineInsights()
        val analytics = CaffeineInsights()

        assertEquals(CaffeineSleepImpactStatus.UNLIKELY, caffeineSleepImpactStatus(home))
        assertTrue(home.entryInsights.isEmpty())
        // With no threshold and a flat curve the scale floors at 1 mg, probed with the smallest curve the chart draws.
        assertEquals(
            1.0,
            caffeineCurveMaxMg(
                points = listOf(
                    CaffeinePoint(time = morning, valueMg = 0.0),
                    CaffeinePoint(time = morning.plusSeconds(3600), valueMg = 0.0),
                ),
                thresholdMg = 0.0,
            ),
            0.0001,
        )
        assertTrue(caffeineDistributionBars(analytics.sourceTotals).isEmpty())
        assertNull(analytics.sourceTotals.firstOrNull()?.label)
    }

    @Test
    fun `the sleep verdict compares bedtime first, then right now`() {
        assertEquals(
            CaffeineSleepImpactStatus.MAY_AFFECT_SLEEP,
            statusFor(currentMg = 120.0, bedtimeMg = 80.0),
        )
        // Over the line now, but back under it by bedtime.
        assertEquals(
            CaffeineSleepImpactStatus.ELEVATED_NOW,
            statusFor(currentMg = 120.0, bedtimeMg = 30.0),
        )
        assertEquals(
            CaffeineSleepImpactStatus.UNLIKELY,
            statusFor(currentMg = 20.0, bedtimeMg = 5.0),
        )
    }

    @Test
    fun `the bedtime card is safe exactly at the threshold`() {
        val insights = CaffeineInsights(bedtimeMg = 50.0, sleepThresholdMg = 50)

        assertTrue(caffeineBedtimeIsSafe(insights))
        // …and the verdict agrees: 50 is not ABOVE 50.
        assertEquals(CaffeineSleepImpactStatus.UNLIKELY, caffeineSleepImpactStatus(insights))
    }

    @Test
    fun `the curve maximum fits the tallest point and the threshold`() {
        val maxMg = caffeineCurveMaxMg(
            points = listOf(
                CaffeinePoint(time = morning, valueMg = 90.0),
                CaffeinePoint(time = morning.plusSeconds(2 * 3600), valueMg = 140.0),
            ),
            thresholdMg = 50.0,
        )

        assertEquals(140.0, maxMg, 0.0001)
    }

    @Test
    fun `a distribution card shows six bars, scaled against their own tallest`() {
        val sourceTotals = listOf(
            slice("Coffee", 400.0),
            slice("Tea", 200.0),
            slice("Cola", 100.0),
            slice("Chocolate", 50.0),
            slice("Energy drink", 40.0),
            slice("Matcha", 30.0),
            // The seventh does not fit on the card.
            slice("Yerba mate", 20.0),
        )

        val bars = caffeineDistributionBars(sourceTotals)
        assertEquals(6, bars.size)
        assertEquals("Coffee", bars.first().label)
        assertEquals(1.0, bars.first().fraction, 0.0001)
        assertEquals(0.5, bars[1].fraction, 0.0001)
        assertEquals("Coffee", sourceTotals.firstOrNull()?.label)

        val buckets = caffeineTimeBucketBars(
            listOf(
                CaffeineTimeBucket(bucket = CaffeineTimeOfDayBucket.MORNING, valueMg = 300.0),
                CaffeineTimeBucket(bucket = CaffeineTimeOfDayBucket.EVENING, valueMg = 150.0),
            )
        )
        assertEquals(CaffeineTimeOfDayBucket.MORNING, buckets.first().bucket)
        assertEquals(1.0, buckets.first().fraction, 0.0001)
        assertEquals(0.5, buckets.last().fraction, 0.0001)
    }

    private fun statusFor(currentMg: Double, bedtimeMg: Double): CaffeineSleepImpactStatus =
        caffeineSleepImpactStatus(
            CaffeineInsights(
                currentMg = currentMg,
                bedtimeMg = bedtimeMg,
                sleepThresholdMg = 50,
                bedtime = LocalTime.of(23, 0),
            )
        )

    private fun slice(label: String, mg: Double): CaffeineDistributionSlice =
        CaffeineDistributionSlice(label = label, valueMg = mg)
}
